Added validation and parsing of the xml part of sxfm

ChristianKaltenecker opened the pull request.

s9latimm committed:
WIP: Add node class

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
A	include/vara/Feature/Group.h
A	include/vara/Feature/Node.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
WIP: Fix bindings

M	bindings/python/vara-feature/pybind_Feature.cpp

Kallistos committed:
constraint tree structure added, builder is missing

Kallistos committed:
Merge branch 'f-FMConstraints' of https://github.com/se-passau/vara-feature into f-FMConstraints

s9latimm committed:
WIP: Refactor feature tree

M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
A	include/vara/Feature/FeatureTreeNode.h
D	include/vara/Feature/Node.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	tools/fm-viewer/FeatureModelViewer.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Clean up, added TODO's

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/Relationship.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp

s9latimm committed:
Make inheritence private

M	include/vara/Feature/FeatureTreeNode.h

s9latimm committed:
Suggested changes

M	.travis.yml
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/OrderedFeatureVector.h
M	include/vara/Feature/Relationship.h
M	lib/Feature/FeatureModel.cpp

Kallistos committed:
Added ConstraintOperator enum, to replace string representation (eases parsing), deleted friends in Constraints

Kallistos committed:
constraint tree, added classes for operands to extend type information of the tree nodes

Commit:
Added validation and parsing of the xml part of sxfm

Commit:
Fix failing test

Commit:
Fix case

Commit:
Fixes to clang-tidy

Kallistos committed:
annotated changes from @vulder in the pull request

Kallistos commented on unittests/resources/test.sxfm:
```diff
@@ -0,0 +1,53 @@
+	:g [1,1] mpm(mpm)
+		: worker(worker)
+		: event(event)
+	:o tls(tls)		:o aesTls(aesTls)			:g [0,1] Group_0(Group_0)
```
are multiple nodes in one line allowed?

ChristianKaltenecker replied:
Nope. Seems to be a copy & paste error. This is fixed in the next revision. Thanks ;-)

Kallistos commented on unittests/resources/sxfm_example.sxfm:
```diff
@@ -0,0 +1,51 @@
+    :r root (root_id)
+      :o opt1 (id_opt1)
+      :o opt2 (id_opt2)
+      :m man1
```
is it allowed that mandatory features have no (display-)name ?

ChristianKaltenecker replied:
If you are referring to the missing ID: yes, this is allowed. 
To be honest, I would simply enforce that the names have to be unique and give each of the features the id 'id_<name>'. This would simplify a lot...
@Kallistos What do you think?

Kallistos replied:
That depends what the names of the features are.
If the feature name is the unique part right before the parenthesis, that is okay for me.
In SPLConqueror this is different right now. There the part in parenthesis is the name that is used as feature name.

ChristianKaltenecker replied:
Correct, the feature name is the unique part right before the parenthesis. 
Yes, you are right regarding SPL Conqueror. However, I do not intend to implement is exactly as it is implemented in SPL Conqueror, as the implementation is partially error-prone.

ChristianKaltenecker replied:
Btw. the mandatory feature above is from the original SXFM example of the SPLOT repository (http://www.splot-research.org/).

Kallistos replied:
Okay!
Then the Id part in the parenthesis seems pretty useless anyways, right?

ChristianKaltenecker replied:
Under the assumption that the feature names are already unique: yes, definitely!

s9latimm committed:
Update FMBuilder with basic constraints

M	.travis.yml
M	CMakeLists.txt
M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/CMakeLists.txt
A	lib/Feature/Constraint.cpp
M	lib/Feature/FeatureModel.cpp

Kallistos committed:
merge conflict solved

Kallistos committed:
renamed XOr (->Xor)

Commit:
Fix test file

s9latimm committed:
Add override

M	include/vara/Feature/Constraint.h

Kallistos committed:
removed unnecassary toString in abstractClasses Binary and Unary Constraint

s9latimm committed:
Add backref

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Merge branch 'f-FMConstraints' of github.com:se-passau/vara-feature into f-FMConstraints

M	include/vara/Feature/Constraint.h

s9latimm committed:
Remove toString

M	include/vara/Feature/Constraint.h

s9latimm committed:
Fix travis

M	.travis.yml

s9latimm committed:
Update variant

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp

s9latimm committed:
Update primay

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp

s9latimm committed:
Make abstract

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h

Kallistos committed:
implemented toString for constraint classes

Kallistos committed:
Merge branch 'f-FMConstraints' of https://github.com/se-passau/vara-feature into f-FMConstraints

s9latimm committed:
Print constraints

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp
M	lib/Feature/FeatureModelParser.cpp

Commit:
Implemented parsing of the feature tree without or groups

s9latimm committed:
Add relationship container

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp

s9latimm committed:
Add tests

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/NumericFeature.cpp
A	unittests/Feature/Relationship.cpp

Kallistos commented on include/vara/Feature/FeatureModelParser.h:
```diff
@@ -90,9 +90,16 @@ class FeatureModelSxfmParser : public FeatureModelParser {
   static bool parseConstraints(xmlChar *Constraints);
 
+  static int countOccurrences(const string& Target, const string& StringToSearch);
+  static std::optional<std::tuple<string, string>> extractCardinality(const string& StringToExtractFrom);
```
why are the cardinalities represented as strings instead of ints, or Integer?

ChristianKaltenecker replied:
Because there is also the '*' sign for the wildcard. Other than that, an integer would suffice.

Kallistos replied:
You could use INT_MAX for '*'

ChristianKaltenecker replied:
I would rather prefer '-1' because '-1' as indentation is not a possible value ;-)
Or, because I have no other negative values, I could also use an unsigned int and use INT_MAX for '*'.

Kallistos commented on include/vara/Feature/FeatureModelParser.h:
```diff
@@ -90,9 +90,16 @@ class FeatureModelSxfmParser : public FeatureModelParser {
+  static string readUntil(const string& StringToReadFrom, const char& CharToSearch, std::string::size_type Start);
+  bool
+  processFeatureTreeLine(std::stringstream& Ss, const char *ParentName,
+                         std::optional<std::tuple<string, string>> ParentCardinality,
```
same as above

Kallistos commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+    bool Opt;
+    int LastIndentation = -1;
+    int OrGroupCounter = 0;
+    std::map<int, string> IndentationToParentMapping;
```
Why do you remember different indentations ? and what happens if you have two indentations that are on the same level but have different parents?

Isn't the current indentation level sufficient?

ChristianKaltenecker replied:
We do have to remember the different indentations for cases such as:
```
:r root (root_id)
  :o opt1 (id_opt1)
    :o opt2 (id_opt2)
      :o opt3 (id_opt3)
  :o opt4 (id_opt4)
```
The indentation between op3 and opt4 is -2 and the parent of opt4 is root and not opt2.
Thus, it does not suffice to only remember the last parent. 

> and what happens if you have two indentations that are on the same level but have different parents?

Should be no problem since a new parent of a certain indentation level overwrites the last one. There is no case where we would still need the last parent.

Kallistos commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+      int CurrentIndentationLevel = countOccurrences(To, Indentation);
+      int Diff = LastIndentation - CurrentIndentationLevel;
+      if ((LastIndentation != -1) && (Diff < -1 | Diff > 1)) {
+        std::cerr << "Indentation error in feature tree." << std::endl;
```
You could add more specific information about the indentation: what you expected and what you got.

ChristianKaltenecker replied:
Sure. I added more comments and also found a small bug in line 279. 
The bug is now also covered in a test.

Kallistos commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<string, string>> Cardinalities;
+      switch (To.at(Pos - 1)) {
```
There is a check for the colon missing.

ChristianKaltenecker replied:
Correct. Will be included in the next commit.

Kallistos commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+        break;
+      }
+      // Extract the name
+      Name = readUntil(To, ' ', Pos + 1);
+
+      // Note that we ignore the ID and use the name of the feature
+      // as unique identifier.
+      if (Name.find_first_of('(') != std::string::npos) {
+        Name = readUntil(Name, '(', 0);
+      }
+
+      // Remove the cardinality
+      if (Name.find_first_of('[') != std::string::npos) {
+        Name = readUntil(Name, '[', 0);
+      }
+
+      // If there is no name, provide an artificial one
+      if (Name.empty()) {
+        OrGroupCounter++;
+        Name = "group_" + std::to_string(OrGroupCounter);
+      }
```
after you checked the indentation you could just split the input line at whitespaces 

`std::vector<std::string> tokens;`
`split(To, " ", std::back_inserter(tokens));`

Kallistos commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+  // Remove the cardinality
+  if (Name.find_first_of('[') != std::string::npos) {
+    Name = readUntil(Name, '[', 0);
+  }
```
That is the same code as at line 277, why did you wrote that twice?

ChristianKaltenecker replied:
Ah, I forgot to remove this could after I changed my approach from a recursive way to a loop.
Will be removed with the next commit.

Kallistos commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+    Result += StringToReadFrom.at(Pos);
+    Pos++;
+  }
+  return Result;
```
you might not need that method if you use the split method on strings

Commit:
Remove unnecessary method

s9latimm committed:
Extend CK

M	include/vara/Feature/Constraint.h

Commit:
Add cardinality parsing and error handling

Commit:
Merge branch f-FMConstraints into sxfm-parser

Commit:
Finish parsing of the feature tree

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+}
+
+/// This method checks if the given feature model is valid
+/// \return true iff the feature model is valid
```
```suggestion
/// This method checks if the given feature model is valid
///
/// \returns true iff the feature model is valid
```

And please move the docs to the header file.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+
+/// This method checks if the given feature model is valid
+/// \return true iff the feature model is valid
+bool FeatureModelSxfmParser::verifyFeatureModel() { return parseDoc().get(); }
```
Could be an inline function in the header.

vulder commented on include/vara/Feature/FeatureModelParser.h:
```diff
@@ -64,6 +64,41 @@ class FeatureModelXmlParser : public FeatureModelParser {
+/// \brief Parsers for feature models in SXFM.
+/// The SXFM (simple XML feature model) format is a feature model format embeded
+/// in an XML structure.
+class FeatureModelSxfmParser : public FeatureModelParser {
```
Could you in general, please try to order the functions in the cpp file as they are in the header file :smile:. Makes it easier to find.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// \param StringToReadFrom string to read from
+/// \param CharToSearch character to search for
+/// \param Start the starting position to start reading
+/// \return the string read until the given character is found
```
```suggestion
/// This method reads beginning from a certain starting position and reads in
/// every character until CharToSearch is found.
///
/// \param StringToReadFrom string to read from
/// \param CharToSearch character to search for
/// \param Start the starting position to start reading
///
/// \returns the string read until the given character is found
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// \param CharToSearch character to search for
+/// \param Start the starting position to start reading
+/// \return the string read until the given character is found
+string FeatureModelSxfmParser::readUntil(const string& StringToReadFrom, const char& CharToSearch, std::string::size_type Start) {
```
You can just use `llvm::StringRef::find`

ChristianKaltenecker replied:
I have done that now. I could also remove another method, which counted the occurrences of a certain string in another one.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// either UINT_MAX for the wildcard or the cardinality number as integer.
+/// \param CardinalityString the cardinality to parse
+/// \return an optional that contains no integer in case of failure or UINT_MAX
+/// for wildcard, or the number itself.
```
```suggestion
/// This method parses the given cardinality and returns an optional.
/// If the optional is empty, the process failed; otherwise the result contains
/// either UINT_MAX for the wildcard or the cardinality number as integer.
///
/// \param CardinalityString the cardinality to parse
///
/// \returns the parsed number, UINT_MAX for wildcards, and nullopt in case of a failure
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// \return an optional that contains no integer in case of failure or UINT_MAX
+/// for wildcard, or the number itself.
+std::optional<int> FeatureModelSxfmParser::parseCardinality(const string& CardinalityString) {
+  std::optional<int> Result = std::optional<int>();
```
```suggestion
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+  if (CardinalityString == "*") {
+    // We use UINT_MAX as our magic integer (which is -1 as int) to indicate
+    // that the cardinality is a wildcard.
+    Result = UINT_MAX;
```
```suggestion
    return UINT_MAX;
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+    }
+  }
+
+  return Result;
```
```suggestion
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+    LongNumber = strtol(CardinalityString.c_str(), &End, 0);
+    if (errno == ERANGE || LongNumber < INT_MIN || LongNumber > INT_MAX
+        || *CardinalityString.c_str() == '\0' || *End != '\0') {
+      std::cerr << "The following cardinality is not integer: " << CardinalityString << std::endl;
```
```suggestion
      llvm::errs() << "The following cardinality is not integer: " << CardinalityString << '\n';
```
We try to use the llvm out and error stream because they are a bit easier to use and have some interesting features. So, to be consistent, it would be good to use them everywhere.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+    if (errno == ERANGE || LongNumber < INT_MIN || LongNumber > INT_MAX
+        || *CardinalityString.c_str() == '\0' || *End != '\0') {
+      std::cerr << "The following cardinality is not integer: " << CardinalityString << std::endl;
+    } else {
```
```suggestion
      return std::nullopt;
    } else {
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// \param CardinalityString the cardinality to parse
+/// \return an optional that contains no integer in case of failure or UINT_MAX
+/// for wildcard, or the number itself.
+std::optional<int> FeatureModelSxfmParser::parseCardinality(const string& CardinalityString) {
```
If you only read a string it is good to only use a `std::string_view`, for llvm there is a similar type called `llvm::StringRef`. Try to use a `StringRef` here, makes it easier because there is already a function `getAsInteger` :smile:

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+        || *CardinalityString.c_str() == '\0' || *End != '\0') {
+      std::cerr << "The following cardinality is not integer: " << CardinalityString << std::endl;
+    } else {
+      Result = LongNumber;
```
```suggestion
      return LongNumber;
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// The cardinality is wrapped in square brackets (e.g., [1,1])
+/// \param StringToExtractFrom the string to extract the cardinality from
+/// \return
+std::optional<std::tuple<int, int>> FeatureModelSxfmParser::extractCardinality(const string& StringToExtractFrom) {
```
```suggestion
std::optional<std::tuple<int, int>> FeatureModelSxfmParser::extractCardinality(const string& StringToExtractFrom) {
```
Can take a StringRef, see description bellow.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+std::optional<std::tuple<int, int>> FeatureModelSxfmParser::extractCardinality(const string& StringToExtractFrom) {
+  std::optional<int> MinCardinality;
+  std::optional<int> MaxCardinality;
+
```
With StringRefs all the positon/string handling becomes much easier.

See bellow

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+  return Dtd;
+}
+
+bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
```
This function is quite huge and should be split into different helper functions.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+    // third a vector consisting of the name of the children
+    std::map<int, std::tuple<string, Relationship::RelationshipKind, std::vector<string>>> OrGroupMapping;
+
+    if (FeatureTree == nullptr) {
```
```suggestion
    if (!FeatureTree) {
```
Could also be

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+      std::optional<std::tuple<int, int>> Cardinalities;
+
+      switch (To.at(Pos - 1)) {
+      case 'o':
+        // Code for optional
```
If these code have special meaning, would it make sense to create a small enum that make the mapping clear? So we could write:
```
case SXFM_CHAR_CODE::OPTIONAL:
    // ... do stuff
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+      // The first character has to be a colon followed by the type of
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<int, int>> Cardinalities;
```
In `cpp` files it can be helpful to add using decl for specific names that are used often and don't collide.
E.g.:
```
using std::optional;

// ... code
```
```suggestion
      optional<std::tuple<int, int>> Cardinalities;
```

PS: there is also using namespace, however, I would recommend to use single using decls because it makes it clearer at the top of the file what names are importet. Furthermore, it reduces the chance of unwanted name clashes, especially if frameworks/stdlib or otherthings are updated.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+      }
+
+      if (LastIndentationLevel != -1) {
+        auto Parent = IndentationToParentMapping.find(CurrentIndentationLevel - 1);
+        assert(Parent != IndentationToParentMapping.end());
```
Hmm, is this more the case that "this should normally never happen" or "a wrong formatted file causes this case". In case of the second one, I would suggest different error handling because if this case happens in an optimized build -- no asserts -- the dereference bellow is UB and could/will crashe hard.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+    }
+
+    // Add the remaining or groups
+    for (auto & OrGroup : OrGroupMapping) {
```
```suggestion
    for (const auto &OrGroup : OrGroupMapping) {
```
Formatting + const.
`OrGroup` is never modified in the loop and can, therefore, be const.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+/// \param StringToSearch the string to search in
+/// \param StringToFind the string to count the occurrences of
+/// \return the number of occurrences
+int FeatureModelSxfmParser::countOccurrences(const string& StringToSearch, const string& StringToFind) {
```
```suggestion
int FeatureModelSxfmParser::countOccurrences(const string& StringToSearch, const string& StringToFind) {
```
Has different parameter names in the header file.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+  return Result;
+}
+
+/// This method counts the occurrences of the second argument in the first argument.
+/// The intended use of this method is to count the indentation level of the feature
+/// tree in the sxfm format.
+/// \param StringToSearch the string to search in
+/// \param StringToFind the string to count the occurrences of
+/// \return the number of occurrences
```
Move to header.

```suggestion
/// This method counts the occurrences of the second argument in the first argument.
/// The intended use of this method is to count the indentation level of the feature
/// tree in the sxfm format.
///
/// \param StringToSearch the string to search in
/// \param StringToFind the string to count the occurrences of
///
/// \returns number of occurrences
```

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+}
+
+bool FeatureModelSxfmParser::parseConstraints(xmlChar *Constraints) {
+  // TODO (CK): This has to wait until the constraint part is implemented in
+  // the FeatureModelBuilder (issue #24)
```
```suggestion
  // TODO (se-passau/VaRA#$NUMBER): This has to wait until the constraint part is implemented in
  // the FeatureModelBuilder
```
The issue number `issue #24` is not in the issue tracker, did you miss typed this?

ChristianKaltenecker replied:
Yes, I misstyped this.I have changed the reference to issue `#702`, which is the right one.

vulder replied:
thx

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+}
+
+bool FeatureModelSxfmParser::parseVm(xmlNode *Node) {
+  // Parse the name first. This procedure is the same as for parsing the
+  // XML file.
```
a) what do you mean by "the same as for parsing the xml file" Is sxfm not also an Xml file?
b) If it is the same we could factor our a method that does the work for both

ChristianKaltenecker replied:
I have adjusted this comment, since there are some differences to the way how xml files are parsed.
I see little reason to refactor here.

vulder replied:
k

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+  // XML file.
+  {
+    std::unique_ptr<xmlChar, void (*)(void *)> Cnt(xmlGetProp(Node, XmlConstants::NAME),
+                                                   xmlFree);
```
Can this fail and initialize `Cnt` with a nullptr?

ChristianKaltenecker replied:
No, this shouldn't happen, since I demand the name attribute to be included in the sxfm model. If it is not available, the xml verification would actually fail **before** it is even parsed and processed.

vulder replied:
k

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+          if (Cnt) {
+            Indentation = reinterpret_cast<char*>(Cnt.get());
+          }
+
```
```suggestion
```

vulder reviewed (changes requested):
I did a first pass over the changes, if you have questions just ping me :smiley:

Commit:
Sort methods; move and format method documentation

Commit:
Adjust reference in comment

Commit:
Use StringRef methods

ChristianKaltenecker commented:
This will be addressed in another PR.
